File: PopulatingNextRightPointersII.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
import java.util.Queue;
import java.util.LinkedList;
public class PopulatingNextRightPointersII {
    public static void main(String[] args) {
        PopulatingNextRightPointersII p = new PopulatingNextRightPointersII();
        TreeLinkNode r = new TreeLinkNode(1);
        r.left = new TreeLinkNode(2);
        p.connect(null);
    }
    public void connect(TreeLinkNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        Queue<TreeLinkNode> q = new LinkedList<TreeLinkNode>();
        Queue<Integer> level = new LinkedList<Integer>();
        q.offer(root);
        level.offer(1);
        TreeLinkNode last = null;
        int levelLast = 1;
        boolean end = true;
        while(q.size() != 0) {
            TreeLinkNode tail = q.poll();
            int tailLevel = level.poll();

            //System.out.println(i);
            if (!end) {
                last.next = tail;
            }
            if (levelLast!=tailLevel) {
                end = true;
                tail.next = null;
            } else {
                end = false;
            }

            last = tail;
            levelLast = tailLevel;

            if (tail != null) {
                if (tail.left != null) {
                    q.offer(tail.left);
                    level.offer(tailLevel+1);
                }
                if (tail.right != null) {
                    q.offer(tail.right);
                    level.offer(tailLevel+1);
                }
            }

        }
    }



}

File: PopulatingNextRightPointersII.cpp
//ac
#include <cassert>
#include <iostream>
#include <vector>
#include <map>
#include <cstdio>
#include <queue>
#include <utility>
#include <algorithm>

using namespace std;
struct TreeLinkNode {
    int val;
    TreeLinkNode *left, *right, *next;
    TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
};


struct Node {
public:
    TreeLinkNode *node;
    int level;
    Node() {
        node = NULL;
        level = -1;
    }
    Node(TreeLinkNode *n, int l) {
        node = n;
        level = l;
    }
};


class Solution {
public:
    void connect(TreeLinkNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        queue<Node> q;
        if (root == NULL)
        {
            return;
        }
        Node r(root, 1);
        q.push(r);

        Node last;
        while(!q.empty()) {
            Node cur = q.front();
            q.pop();
            int lev = cur.level;
            if (cur.node->left != NULL)
            {
                Node left(cur.node->left, lev+1);
                q.push(left);
            }
            if (cur.node->right != NULL)
            {
                Node right(cur.node->right, lev+1);
                q.push(right);
            }
            if (last.node != NULL)
            {
                if (last.level == cur.level)
                {
                    last.node->next = cur.node;
                }
                else
                {
                    last.node->next = NULL;
                }
            }
            last = cur;
            if(q.empty())
            {
                cur.node->next = NULL;
            }
        }
    }
};

int main() {
    Solution s;
    TreeLinkNode *n(0);
    s.connect(n);
    return 0;
}